reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/RepeatUntil.kt

```diff
@@ -0,0 +1,44 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.ErrorCallback
+import com.badoo.reaktive.base.ValueCallback
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.serializer.serializer
+
+fun <T> Observable<T>.repeatUntil(predicate: () -> Boolean): Observable<T> =
+    observable { emitter ->
+        val observer =
+            object : ObservableObserver<T>, ValueCallback<T> by emitter, ErrorCallback by emitter {
+                // Prevents recursive subscriptions
+                private val serializer =
+                    serializer<Unit> {
+                        subscribe(this)
+                        true
+                    }
+
+                override fun onSubscribe(disposable: Disposable) {
+                    emitter.setDisposable(disposable)
+                }
+
+                override fun onComplete() {
+                    emitter.tryCatch(
+                        block = { predicate() },
+                        onSuccess = {
+                            if (!emitter.isDisposed) {
```

Comment:
Will do that, but passing predicate without invoking it will return lambda on success so I will have to do something like 
`emitter.tryCatch(
                        block = { predicate },
                        onSuccess = { shouldRepeat ->
                            if (shouldRepeat()) {
                                emitter.onComplete()
                            } else if (!emitter.isDisposed) {
                                subscribeToUpstream()
                            }
                        }
                    )`
So better send the predicate as { predicate(value) } to get a boolean in onSuccess. Since the lambda expected in block is different than the predicate not like the Observable.repeatUntil

What do you think?

Reply:
tryCatch accepts a function to be invoked. For Observable case predicate has same type so can be passed directly. For Single case the types are different and if you pass directly it won't compile. So it's makes sense only for Observable.

Reply:
Cool, for Observable I am directly passing the predicate but for Single I am passing `{ predicate(value) }`. Check `Singe.RepeatUntil` please.